tests/base/quadrature_point_data_04.cc

```diff
@@ -0,0 +1,247 @@
+              data_vec[i]->pt = fe_values.quadrature_point(i);
+          }
+      }
+}
+
+DeclException3(ExcWrongValue,
+               double,
+               double,
+               double,
+               << "Received " << arg1 << ", Expected " << arg2
+               << ", delta = " << arg3);
+
+void
+check_data(const Triangulation<2> & tria,
+           FEValues<2> &            fe_values,
+           const QuadratureStorage &storage)
+{
+  deallog << "Checking quadrature cell data" << std::endl;
+  constexpr double eps = 1e-10;
+  for (auto &cell : tria.active_cell_iterators())
+    if (cell->is_locally_owned())
+      {
+        fe_values.reinit(cell);
+        if (cell->material_id() == 0)
+          {
+            const auto data_vec = storage.template get_data<Mat0>(cell);
+            for (unsigned int i = 0; i < n_data_points_per_cell; ++i)
+              {
+                const double value         = data_vec[i]->x;
+                const double correct_value = fe_values.quadrature_point(i)[0];
+                AssertThrow(std::fabs(value - correct_value) < eps,
+                            ExcWrongValue(value,
+                                          correct_value,
+                                          value - correct_value));
+              }
+          }
+        else if (cell->material_id() == 1)
+          {
+            auto data_vec = storage.template get_data<Mat1>(cell);
+            for (unsigned int i = 0; i < n_data_points_per_cell; ++i)
+              {
+                const double value =
+                  fe_values.quadrature_point(i).distance(data_vec[i]->pt);
+                const double correct_value = 0.;
+                AssertThrow(std::fabs(value - correct_value) < eps,
+                            ExcWrongValue(value,
+                                          correct_value,
+                                          value - correct_value));
+              }
+          }
+      }
+}
+
+int
+main(int argc, char *argv[])
+{
+  Utilities::MPI::MPI_InitFinalize mpi_initialization(argc, argv, 1);
+  mpi_initlog();
+
+  // create a mesh with four cells
+  parallel::distributed::Triangulation<2> tria(MPI_COMM_WORLD);
+  GridGenerator::subdivided_hyper_cube(tria, 2);
+
+  // assign material ids
+  for (auto &cell : tria.active_cell_iterators())
+    {
+      const auto center = cell->center();
+      if (center[0] < 0.5 && center[1] < 0.5)
+        cell->recursively_set_material_id(0);
+      else if (center[0] >= 0.5 && center[1] < 0.5)
+        cell->recursively_set_material_id(1);
+      else if (center[0] < 0.5 && center[1] >= 0.5)
+        cell->recursively_set_material_id(2);
+      else
+        // material_id 3 does not have an associated data structure
+        cell->recursively_set_material_id(3);
+    }
+
+  // create a history structure and populate it
+  QuadratureStorage storage;
+  initialize_data(tria, storage);
+
+  // Assign some value to the cell data
+  FE_Q<2>     fe(1);
```

Comment:
Why is `degree` not  a constexpr, but `n_quadrature_per_dim`?

Reply:
I removed the consexpr global variable and made it more consistent.